src/proto/Cosmos.proto

```diff
@@ -205,6 +205,62 @@ message Message {
         bytes auth_info_bytes = 2; 
     }
 
+    // StakeAuthorization defines authorization for delegate/undelegate/redelegate.
+    //
+    // Since: cosmos-sdk 0.43
+    message StakeAuthorization {
+        // max_tokens specifies the maximum amount of tokens can be delegate to a validator. If it is
+        // empty, there is no spend limit and any amount of coins can be delegated.
+        Amount max_tokens = 1;
+        // validators is the oneof that represents either allow_list or deny_list
+        oneof validators {
+            // allow_list specifies list of validator addresses to whom grantee can delegate tokens on behalf of granter's
+            // account.
+            Validators allow_list = 2;
+            // deny_list specifies list of validator addresses to whom grantee can not delegate tokens.
+            Validators deny_list = 3;
+        }
+        // Validators defines list of validator addresses.
+        message Validators {
+            repeated string address = 1;
+        }
+        // authorization_type defines one of AuthorizationType.
+        AuthorizationType authorization_type = 4;
+    }
+
+    // AuthorizationType defines the type of staking module authorization type
+    //
+    // Since: cosmos-sdk 0.43
+    enum AuthorizationType {
+        // AUTHORIZATION_TYPE_UNSPECIFIED specifies an unknown authorization type
+        AUTHORIZATION_TYPE_UNSPECIFIED = 0;
+        // AUTHORIZATION_TYPE_DELEGATE defines an authorization type for Msg/Delegate
+        AUTHORIZATION_TYPE_DELEGATE = 1;
+        // AUTHORIZATION_TYPE_UNDELEGATE defines an authorization type for Msg/Undelegate
+        AUTHORIZATION_TYPE_UNDELEGATE = 2;
+        // AUTHORIZATION_TYPE_REDELEGATE defines an authorization type for Msg/BeginRedelegate
+        AUTHORIZATION_TYPE_REDELEGATE = 3;
+    }
+
+
+    // cosmos-sdk/MsgGrant
+    message AuthGrant {
+        string granter = 1;
+        string grantee = 2;
+        oneof grant_type {
+            StakeAuthorization grant_stake = 3;
+        }
+        int64 expiration = 4;
+        string type_url = 5;
```

Comment:
can we hide this for the app / caller?

Reply:
The end user have to choose an expiration date for his authz:

<img width="810" alt="Screenshot 2022-08-21 at 20 38 31" src="https://user-images.githubusercontent.com/21139416/185805907-7bdd49bf-d0fd-4085-a713-84d75c206367.png">

`type_url` and `one_of` are also required because the **authz** module can provide authorization for multiple things:

- Bank authz
- Delegate authz
- Fee authz 

If we want to support only compounding, then yes I can hide `type_url`, but I think currently it makes the code more flexible for future additions.

Reply:
For example, you know `grant_type`, can you infer the type url?, the app needs to define `/cosmos.staking.v1beta1.StakeAuthorization` somewhere too
```swift
$0.typeURL = "/cosmos.staking.v1beta1.StakeAuthorization"
$0.grantStake = CosmosMessage.StakeAuthorization.with {
    $0.allowList.address = ["cosmosvaloper1gjtvly9lel6zskvwtvlg5vhwpu9c9waw7sxzwx"]
    $0.authorizationType = CosmosMessage.AuthorizationType.delegate
}
```

Reply:
Yes this is possible!